bokehjs/src/lib/models/glyphs/webgl/dash_cache.ts

```diff
@@ -0,0 +1,167 @@
+    const dtex = 0.5*twice_jumps_gcd * ideal_ntex/ntex
+
+    // xstart is the position along the texture of the first value, and offset
+    // is the distance to the upstroke of the first dash.
+    // When interpolating the texture each texel fills 1/ntex of the length of
+    // the texture.  For a single dash the centre of the dash is 0.25 along
+    // the texture, so the upstroke offset has to be determined from this.
+    let xstart
+    if (length_pow_2) {
+      xstart = 0.5*pattern[0]
+
+      if (dtex < xstart) {
+        const n_dtex = Math.floor(xstart / dtex)
+        xstart -= n_dtex*dtex
+      }
+    } else {
+      // Have lots of values so don't need to match middles of dashes/gaps.
+      xstart = 0.0
+    }
+    const offset = xstart - 0.5*dtex
+
+    // Calculate values for texture.
+    const dist = new Uint8Array(ntex)
+    let dash_index = 0
+    for (let i = 0; i < ntex; i++) {
+      const x = xstart + i*dtex  // Distance along texture.
+
+      // Which dash are we in?
+      if (x > starts_and_ends[dash_index + 1])
+        dash_index++
+
+      const xsize = pattern[dash_index]
+      const xmid = starts_and_ends[dash_index] + 0.5*xsize
+      let dist_float = 0.5*xsize - Math.abs(x - xmid)
+      if (dash_index % 2 == 1)
+        dist_float = -dist_float  // Change sign for gaps between dashes.
+
+      dist[i] = Math.round(255*(dist_float-dist_min) / (dist_max-dist_min))
+    }
+
+    // Create the 1D texture.
+    const tex: Texture2D = this._regl.texture({
+      shape: [ntex, 1, 1],
+      data: dist,
+      wrapS: 'repeat',
+      format: 'alpha',
+      type: 'uint8',
+      mag: 'linear',
+      min: 'linear',
+    })
+
+    return [[len, offset, dist_min, dist_max], tex]
+  }
+
+  protected _get_key(pattern: number[]): string {
+    return pattern.join(",")
+  }
+
+  public _get_or_create(pattern: number[]): DashReturn {
+    const key = this._get_key(pattern)
+    let cached = this._map.get(key)
+    if (cached === undefined) {
+      const scale: number = gcd(pattern)
+
+      if (scale > 1) {
+        // Do not modify pattern in-place, create a new one.
+        pattern = map(pattern, (n) => (n / scale))
+
+        // Get the simple pattern that can be reused when scaled up.
+        cached = this._get_or_create(pattern)
+
+        // Store an entry for the requested pattern which is just the simple
+        // pattern scaled-up.
+        const [tex_info, tex, _simple_scale] = cached
+        cached = [tex_info, tex, scale]
+        this._map.set(key, cached)
+      } else {
+        // Simple pattern with scale of 1.
+        const [tex_info, tex] = this._create_texture(pattern)
+
+        // Store the simple pattern.
+        cached = [tex_info, tex, scale]
+        this._map.set(key, cached)
+      }
+    }
+
+    return cached
+  }
+
+  public get(pattern: number[]): DashReturn {
+    // Limit pattern to even number of items.
+    if (pattern.length % 2 == 1)
+      pattern = pattern.slice(0, -1)
```

Comment:
For the sake of compatibility with 2d canvas, this should be `pattern = concat(pattern, pattern)`. Though it may not happen at all, because odd patterns should be normalized at an earlier stage, thus a simple assertion would also suffice.